src/contract.ts

```diff
@@ -0,0 +1,381 @@
+   * @returns
+   */
+  static async get(
+    program: Program,
+    abiData: string,
+    contractStorageAccount: Keypair
+  ): Promise<Contract> {
+    return new Contract(program, contractStorageAccount, abiData);
+  }
+
+  public abi: ethers.utils.Interface;
+  readonly functions: { [name: string]: ContractFunction };
+
+  /**
+   *
+   * @param program
+   * @param contractStorageAccount
+   * @param abiData
+   */
+  constructor(
+    public program: Program,
+    public contractStorageAccount: Keypair,
+    public abiData: string
+  ) {
+    this.abi = new ethers.utils.Interface(abiData);
+    this.functions = {};
+    Object.entries(this.abi.functions).forEach(([, frag]) => {
+      this.functions[frag.name] = this.buildCall(frag);
+    });
+  }
+
+  /**
+   *
+   * @param fragment
+   * @returns
+   */
+  buildCall(fragment: ethers.utils.Fragment): ContractFunction {
+    return (...args: Array<any>) => {
+      const last = args[args.length - 1];
+      if (args.length > fragment.inputs.length && typeof last === 'object') {
+        return this.call(
+          fragment.name,
+          args.slice(0, fragment.inputs.length),
+          last
+        );
+      } else {
+        return this.call(fragment.name, args);
+      }
+    };
+  }
+
+  /**
+   *
+   * @param name
+   * @param params
+   * @param accounts
+   * @param writableAccounts
+   * @param seeds
+   * @param signers
+   * @param caller
+   * @returns
+   */
+  private async call(
+    name: string,
+    args: any[],
+    options?: ContractTransactionOptions
+  ): Promise<ContractCallResult> {
+    const {
+      accounts = [],
+      writableAccounts = [],
+      seeds = [],
+      signers = [],
+      caller = this.program.payerAccount.publicKey,
+      value = 0,
+      simulate = false,
+    } = options ?? {};
+
+    const fragment = this.abi.getFunction(name);
+    const input = this.abi.encodeFunctionData(name, args);
+
+    const data = Buffer.concat([
+      this.contractStorageAccount.publicKey.toBuffer(),
+      caller.toBuffer(),
+      Buffer.from(numToPaddedHex(value), 'hex'),
+      Buffer.from('00000000', 'hex'),
+      encodeSeeds(seeds),
+      Buffer.from(input.replace('0x', ''), 'hex'),
+    ]);
+
+    const keys = [
+      ...seeds.map((seed) => ({
+        pubkey: seed.address,
+        isSigner: false,
+        isWritable: true,
+      })),
+      {
+        pubkey: this.contractStorageAccount.publicKey,
+        isSigner: false,
+        isWritable: true,
+      },
+      {
+        pubkey: SYSVAR_CLOCK_PUBKEY,
+        isSigner: false,
+        isWritable: false,
+      },
+      {
+        pubkey: PublicKey.default,
+        isSigner: false,
+        isWritable: false,
+      },
+      ...accounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: false,
+      })),
+      ...writableAccounts.map((pubkey) => ({
+        pubkey,
+        isSigner: false,
+        isWritable: true,
+      })),
```

Comment:
This is something I don't know about programs compiled with Solang. Are the extra accounts always guaranteed to be in this order?

Reply:
Solang doesn't care about the order of the accounts.

Reply:
@seanyoung  It actually does.  For example, for some reason, moving these to the end of the list results in an error https://github.com/vbstreetz/solana-solidity.js/blob/master/src/contract.ts#L256-L260.